tracer/src/Datadog.Trace/Ci/GitCommandHelper.cs

```diff
@@ -188,6 +189,222 @@ static List<FileCoverageInfo> ParseGitDiff(string diffOutput)
+                        break;
+                    }
+                }
+            }
+
+            // 4. Build candidate list
+            var branchesOutput = RunGitCommand(
+                workingDirectory,
+                $"for-each-ref --format='%(refname:short)' refs/heads refs/remotes/{remoteName}",
+                MetricTags.CIVisibilityCommands.BuildCandidateList);
+
+            if (branchesOutput is not { ExitCode: 0 } || string.IsNullOrWhiteSpace(branchesOutput?.Output))
+            {
+                Log.Warning("GitCommandHelper: Failed to get branch list");
+                return null;
+            }
+
+            var regex = new Regex(branchFilterPattern);
+            var branches = SplitLines(branchesOutput!.Output)
+                          .Select(b => b.Trim('\'', ' '))
+                          .Where(b => !string.Equals(b, targetBranch, StringComparison.OrdinalIgnoreCase))
+                          .Where(b =>
+                           {
+                               string nameToCheck = b;
+                               if (b.StartsWith($"{remoteName}/"))
+                               {
+                                   nameToCheck = b.Substring(remoteName.Length + 1);
+                               }
+
+                               return regex.IsMatch(nameToCheck);
+                           })
+                          .ToList();
+
+            if (branches.Count == 0)
+            {
+                Log.Warning("GitCommandHelper: No candidate branches found");
+                return null;
+            }
+
+            // 5. Compute metrics for each branch
+            var metrics = new List<Tuple<string, string, int, int>>(); // Branch, MergeBase, Behind, Ahead
+
+            foreach (var branch in branches)
+            {
+                // Find merge-base (common ancestor)
+                var mergeBaseOutput = RunGitCommand(
+                    workingDirectory,
+                    $"merge-base {branch} {targetBranch}",
+                    MetricTags.CIVisibilityCommands.MergeBase);
+
+                if (mergeBaseOutput is not { ExitCode: 0 } || string.IsNullOrWhiteSpace(mergeBaseOutput?.Output))
+                {
+                    continue; // Skip if no common history
+                }
+
+                var mergeBaseSha = mergeBaseOutput!.Output.Trim();
+
+                // Count commits ahead and behind
+                var revListOutput = RunGitCommand(
+                    workingDirectory,
+                    $"rev-list --left-right --count {branch}...{targetBranch}",
+                    MetricTags.CIVisibilityCommands.RevList);
+
+                if (revListOutput is not { ExitCode: 0 } || string.IsNullOrWhiteSpace(revListOutput?.Output))
+                {
+                    continue; // Skip if it cannot get commit counts
+                }
+
+                var counts = revListOutput!.Output.Split(new[] { '\t', ' ' }, StringSplitOptions.RemoveEmptyEntries);
```

Comment:
Array declaration inside the foreach, you can declare it as internal static and avoid this memory allocation in each loop

Reply:
Changing it!